### PR TITLE
✨ Provide a truly lazy restmapper

### PR DESCRIPTION
**Pull request by Fedosin**
This PR adds a rest mapper that will lazily query the provided client for discovery information to do REST mappings. This allows you to significantly reduce the number of requests to the API server, as currently we do a request for each registered GroupVersion at startup. It also makes the startup time of the controller more predictable and independent of the number of CRDs installed in the system.

Fixes: https://github.com/kubernetes-sigs/controller-runtime/issues/1603

### CODE REVIEW

File: pkg/client/apiutil/dynamicrestmapper.go [hunk 1]
```diff
@@ -76,30 +73,43 @@ func WithCustomMapper(newMapper func() (meta.RESTMapper, error)) DynamicRESTMapp
...
+				groupResources, err := restmapper.GetAPIGroupResources(client)
+				if err != nil {
+					return nil, err
+				}
+
+				return restmapper.NewDiscoveryRESTMapper(groupResources), nil
+			}
+		} else {
+			drm.newMapper = func() (meta.RESTMapper, error) {
+				return restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(client)), nil
```
Comment (alvaroaleman) [inline pkg/client/apiutil/dynamicrestmapper.go:104]: From what I can tell this only defers until the first usage and then fetches all groups, rather than the ones that are needed:
1. `RESTMapping` constructs a `delegate`
2. `getDelegate` calls `GetAPIGroupResources`
3. `GetAPIGroupResources` fetches all apis

Let me know if I missed something.

File: pkg/client/apiutil/lazyrestmapper.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	return &LazyRESTMapper{
+		mapper: restmapper.NewDiscoveryRESTMapper([]*restmapper.APIGroupResources{}),
+		client: dc,
+	}, nil
+}
+
+func (m *LazyRESTMapper) addKnownGroupAndReload(groupName string) error {
+	m.knownGroups = append(m.knownGroups, groupName)
```
Comment (JoelSpeed) [inline pkg/client/apiutil/lazyrestmapper.go:53]: Should we not add to `knownGroups` after we have attempted to fetch the group? If it fails, we are calling it a known group even though it's currently unknown to the mapper

File: pkg/client/apiutil/lazyrestmapper.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+	groupResources, err := m.getFilteredAPIGroupResources(m.client)
+	if err != nil {
+		return err
+	}
+
+	m.mapper = restmapper.NewDiscoveryRESTMapper(groupResources)
```
Comment (JoelSpeed) [inline pkg/client/apiutil/lazyrestmapper.go:60]: This looks ripe for a race condition, I think you need a RWMutex around the `m.mapper`

File: pkg/client/apiutil/lazyrestmapper.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// LazyRESTMapper is a RESTMapper that will lazily query the provided
+// client for discovery information to do REST mappings.
+type LazyRESTMapper struct {
+	mapper      meta.RESTMapper
+	client      *discovery.DiscoveryClient
+	knownGroups []string
```
Comment (JoelSpeed) [inline pkg/client/apiutil/lazyrestmapper.go:36]: You could use a `sets.String` for this, would save you some effort around checking the content of the list

File: pkg/client/apiutil/lazyrestmapper.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	return &LazyRESTMapper{
+		mapper: restmapper.NewDiscoveryRESTMapper([]*restmapper.APIGroupResources{}),
+		client: dc,
+	}, nil
+}
+
+func (m *LazyRESTMapper) addKnownGroupAndReload(groupName string) error {
+	m.knownGroups = append(m.knownGroups, groupName)
+
+	groupResources, err := m.getFilteredAPIGroupResources(m.client)
```
Comment (JoelSpeed) [inline pkg/client/apiutil/lazyrestmapper.go:55]: Why is this method a member of the `LazyRestMapper` when you're also passing an argument including one of the members from the `LazyRestMapper`, I would suggest changing this up slightly

```
groups := sets.NewString(m.knownGroups.List()...)
groups.Add(groupName)

groupResources, err := getFilteredAPIGroupResources(m.client, groups)
handle(err)

m.knownGroups.Add(groupName)
```

This will make it so we only update the known groups on a successful filter, it also decouples the function from the restmapper struct which makes it less likely to start mutating its state and causing data races.

File: pkg/client/apiutil/lazyrestmapper.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	sgs, err := d.ServerGroups()
+	if sgs == nil {
+		return nil, nil, err
+	}
+	resultGroups := []*metav1.APIGroup{}
+	for i := range sgs.Groups {
+		if m.isGroupKnown((&sgs.Groups[i]).Name) {
+			resultGroups = append(resultGroups, &sgs.Groups[i])
+		}
+	}
```
Comment (JoelSpeed) [inline pkg/client/apiutil/lazyrestmapper.go:199]: Every time we look for a new group, we are also going to discover the old groups as well, is that desirable? Or should we convert the logic to merge the old results with the newly requested group? In that case, if you see an object that isn't mapped, but in a known group, that group will become the "new group" and we will refresh it, but leave the others alone, which will minimise the API calls we make

**Review by JoelSpeed — COMMENTED** [review]
I think this is a worthwile improvement to the controller-runtime project, with large numbers of CRDs, we've seen rest mapping take a long time and cache large amounts of data that isn't really necessary.

I think there are a few improvements that could be made to make this a little cleaner, and one logical optimisation, see my comments and let me know what you think

File: pkg/client/apiutil/lazyrestmapper.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+		return resultGroups, result, nil
+	}
+
+	return resultGroups, result, &discovery.ErrGroupDiscoveryFailed{Groups: failedGroups}
+}
+
+// getFilteredAPIGroupResources uses the provided discovery client to gather
+// discovery information and populate a slice of APIGroupResources.
+func (m *LazyRESTMapper) getFilteredAPIGroupResources(cl discovery.DiscoveryInterface) ([]*restmapper.APIGroupResources, error) {
+	gs, rs, err := m.filteredServerGroupsAndResources(cl)
```
Comment (lobziik) [inline pkg/client/apiutil/lazyrestmapper.go:224]: I think there is no need to pipe down discovery client as an argument here.

Could `m.client` be used in `filteredServerGroupsAndResources`?

**Review by lobziik — COMMENTED** [review]
I mainly agree with @JoelSpeed comments. Have just an extra nit about docstrings below.

File: pkg/client/apiutil/lazyrestmapper.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+			return res, err
+		}
+		res, err = m.mapper.ResourcesFor(input)
+	}
+	return res, err
+}
+
+// RESTMapping implements Mapper.RESTMapping
+func (m *LazyRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*meta.RESTMapping, error) {
+	res, err := m.mapper.RESTMapping(gk, versions...)
```
Comment (alvaroaleman) [inline pkg/client/apiutil/lazyrestmapper.go:115]: This method on the discovery restmapper fetches all api groups, which means this change doesn't make anything more lazy than it is today. Before we merge anything in this area I expect an integration test that verifies that we do exactly one api call for discovery at the time we first use a given group. And I recommend you write that test first to avoid going down more avenues that look like they would solve the problem but don't.

File: pkg/client/apiutil/lazyrestmapper.go [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+		if err = m.addKnownGroupAndReload(input.Group); err != nil {
+			return res, err
+		}
+		res, err = m.mapper.ResourcesFor(input)
+	}
+	return res, err
+}
+
+// RESTMapping implements Mapper.RESTMapping
+func (m *LazyRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*meta.RESTMapping, error) {
```
Comment (alvaroaleman) [inline pkg/client/apiutil/lazyrestmapper.go:114]: I [fiddled around with this some time ago](https://github.com/kubernetes-sigs/controller-runtime/compare/master...alvaroaleman:controller-runtime:lazy-restmapper?expand=1) as well, and learned that:
1. `RESTMapping` when called with a fully qualified groupKind and exactly one version can be implemented with one api call
2. `RestMapping` with the above is the only way controller-runtime itself uses the RESTMapper
3. The upstream restmappers have a bunch of arcane logic to provide something useful even providing zero versions, more than one version, only a prefix to the group rather than the full group and probably more that I am forgetting now

Point one and two mean that this is relatively straight-forward to implement. We can also make the reloading when a given kind is not found more efficient by discarding only the given group+version rather than all restmappings.

Point three complicates things. I honestly have no idea if anyone who uses the RESTMapper uses other methods on it and expects them to work the way they do in upstream. With that POC change of mine I went the route of initializing a normal discovery RESTMapper if someone calls a method other than `RESTMapping` in the way controller-runtime does.

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+		knownGroups: map[string]*restmapper.APIGroupResources{},
+	}, nil
+}
+
+func (m *LazyRESTMapper) addKnownGroupAndReload(groupName string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	groupResources, err := getFilteredAPIGroupResources(m.client, groupName)
+	if err != nil || groupResources == nil {
```
Comment (JoelSpeed) [inline pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:61]: Do we want to handle `groupResources == nil` in a different way? What's the effect of returning a nil here? Will we bubble up a no match error in the end?

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+	if sgs == nil {
+		return nil, nil, err
+	}
+
+	apiGroup := metav1.APIGroup{}
+	for i := range sgs.Groups {
+		if groupName == (&sgs.Groups[i]).Name {
+			apiGroup = sgs.Groups[i]
+		}
+	}
```
Comment (JoelSpeed) [inline pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:206]: What happens if there is no match?

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper_test.go [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+		g.Expect(crt.GetRequestCount()).To(gmg.Equal(3))
+	})
+
+	t.Run("LazyRESTMapper should fetch data based on the request", func(t *testing.T) {
+		g := gmg.NewWithT(t)
+
+		// For each new group mapper performs 3 requests to the API server:
+		// #1: GET https://host/api
+		// #2: GET https://host/apis
+		// #3: GET https://host/apis/<group>/<version>
```
Comment (alvaroaleman) [inline pkg/client/apiutil/lazyrestmapper/lazyrestmapper_test.go:102]: Couple of questions:
1. This isn't actually tested (The test for `getFilteredAPIGroupResources` is insufficient, my reading of the discovery restmapper is that it will do these again) 
2. I don't understand why we need to do step 1. and 2. We know both the group and the version, so only step 3 should be enough
3. If I understand it correctly, as soon as we discover a second group we end up re-doing all these api calls plus the discovery call needed for the second group, rather than just the discovery call for the second group.

Given that this is a stop-gap change until aggregated discovery is available for everyone, I do not want to merge something that doesn't fully solve the original problem, it'll just lead to a new issue of `Provide a truly, truly lazy restmapper`

Comment (Fedosin) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper_test.go]: Hey! Thank you for reviewing the code. I've optimized the mapper significantly, so now it is `The Truly Lazy Restmapper`.

Now, it needs 2 api calls (/api and /apis) to initialize + just one call for each requested api group version. I've added more tests to validate this behavior + refactored the code to make it simpler.

Let me know if there is something else I could improve.

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// LazyRESTMapper is a RESTMapper that will lazily query the provided
+// client for discovery information to do REST mappings.
+type LazyRESTMapper struct {
```
Comment (vincepri) [inline pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: Curious if we though actually about making this RESTMapper the default implementation, truthfully it sounds like that the benefit of making less calls, and only making them when needed are something that a lot of users might want. 

If we decide to go down this path, we should however provide a feature gate of sorts that enables this as an experiment first (opt in) -> on by default -> then enable it and remove the gate.

Thoughts?

Comment (Fedosin) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: Hey! Initially I didn't plan to make it default but rather to use it either for clusters with a slow API server, or for very crd-pumped ones (ahem, OpenShift), when we have to do more than a hundred requests to initialize the mapper.

But after multiple optimizations I could say there is no worsening relative to the current implementation, only improvements. So, it's reasonable to make it default after some testing.

Do you have an example of such feature gating in this project, so I can mark this feature as experimental?

Comment (Fedosin) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: btw, we test this implementation with OpenShift here: https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/213

Comment (vincepri) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: We'd need to introduce a feature gate, which we currently don't have support for. For this PR, we can focus immediately to do some sort of split-logic between how things were done in the default RESTMapper, and this way.

If we feel comfortable though and have enough testing coverage all around, we can mark this a breaking change (⚠️) and add extensive release notes and ask for feedback.

@alvaroaleman wdyt?

Comment (alvaroaleman) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: I'd suggest to simply add this restmapper for now and backport it into 0.14. 0.15 onwards (or later, depending on if we find issues) we can simply make it the default but leave the option to provide a custom one.

**Comment by alvaroaleman** [comment]
Sorry, this is on my queue but I am a bit swamped right now

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go (hunk 13 above)
Comment (vincepri) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: We can definitely backport, although for 0.15 (next version) I'd prefer to have a single RESTMapper, and merge the current Dynamic with the Lazy implementation. We've only ever exposed a single one, and I worry that more than one RESTMapper implementation might create even more confusion than we have today. 

The other option I see is to just merge this as net-new addition to the DynamicRESTMapper implementation, make it default in 0.15 and wait for feedback. 

The current DynamicRESTMapper implementation godoc reads:
```
dynamicRESTMapper is a RESTMapper that dynamically discovers resource types at runtime.
```
which makes it sound like it's dynamic and lazy, an on-demand type rest mapper, but it's really not.

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+		return fmt.Errorf("failed to get API group resources: %w", err)
+	}
+
+	groupResources := &restmapper.APIGroupResources{
+		Group:              apiGroup,
+		VersionedResources: make(map[string][]metav1.APIResource),
+	}
+	for version, resources := range groupVersionResources {
+		groupResources.VersionedResources[version.Version] = resources.APIResources
+	}
```
Comment (alvaroaleman) [inline pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:182]: Can we add a fastpath here for how controller-runtime and 99% of everyone who isn't `kubectl` uses this please:

```suggestion
	apiGroup := metav1.APIGroup{
		Name: groupName,
	}
	for _, version := range versions {
		apiGroup.Versions = append(apiGroup.Versions, metav1.GroupVersionForDiscovery{
			GroupVersion: metav1.GroupVersion{Group: groupName, Version: version}.String(),
			Version:      version,
		})
	}
	groupResources := &restmapper.APIGroupResources{
		Group:              apiGroup,
		VersionedResources: make(map[string][]metav1.APIResource),
	}

	if len(versions) == 1 {
		gv := metav1.GroupVersion{
			Group:   groupName,
			Version: versions[0],
		}
		resources, err := m.client.ServerResourcesForGroupVersion(gv.String())
		if err != nil {
			return fmt.Errorf("failed to get Resources for GroupVersion %s: %w", gv.String(), err)
		}
		groupResources.VersionedResources[versions[0]] = resources.APIResources
	} else {
		// First, find information about requested group and its versions. Fail immediately if there is no such group.
		apiGroup, err := m.findAPIGroupByName(groupName)
		if err != nil {
			return err
		}
		groupResources.Group = apiGroup

		// If no specific versions are set by user, we will scan all available ones for the API group.
		if len(versions) == 0 {
			for _, version := range apiGroup.Versions {
				versions = append(versions, version.Version)
			}
		}

		// Second, get resources. The number of API calls is equal to the number of versions: /apis/<group>/<version>.
		groupVersionResources, err := m.fetchGroupVersionResources(apiGroup.Name, versions...)
		if err != nil {
			return fmt.Errorf("failed to get API group resources: %w", err)
		}

		for version, resources := range groupVersionResources {
			groupResources.VersionedResources[version.Version] = resources.APIResources
		}
	}
```

This gets the number of discovery calls down to one per groupVersion.

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go (hunk 13 above)
Comment (alvaroaleman) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: Merging the two for next release sounds good to me.

I _think_ the new lazy restmapper does everything the current dynamic one does btw. The current dynamic one will on a resource not found error reload the entire restmapping (with some ratelimiting). Since the new one will simply try to load the restmapping on-demand if it doesn't have it cached, that should solve the same problem just as well.

**Review by alvaroaleman — COMMENTED** [review]
@Fedosin one comment to optimize this even further, other than that this looks good to me. Thanks a ton for your work and sorry it took so long.

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go (hunk 13 above)
Comment (vincepri) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: Could we have the implementation of this PR inlined with the current one, but all of it behind a feature gate?

Comment (alvaroaleman) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: How would you featuregate this?

Comment (vincepri) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: A new `DynamicRESTMapperOption` that sets the flag could work?

Comment (vincepri) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: Something like: WithExperimentalLazyMapper

Comment (alvaroaleman) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go:33]: sgtm

File: pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go (hunk 14 above)
Comment (Fedosin) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go]: Hey! Thank you for reviewing my code! I was on PTO, so couldn't update it earlier...

I applied the optimization you suggested and it works even better now! I did it slightly differently to keep the code clean and avoid unnecessary conditions. But everything else works as expected - we always do just one API request for each call. Also I added more tests to cover some corner cases.

The changes I introduced in the recent update (besides the tests) affect `addKnownGroupAndReload` only.

Comment (alvaroaleman) [reply pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go]: Awesome, this looks great, thanks so much!

The only thing left at this point from my POV is the point from the other discussion to not have this in its own package but instead construct this in `apiutil.NewDynamicRestMapper` when a new option is passed as Vince suggested here: https://github.com/kubernetes-sigs/controller-runtime/pull/2116#discussion_r1091137091

If you don't want to do that, we can also merge this as-is and someone else will take care of doing that in a follow-up.

**Comment by Fedosin** [comment]
@alvaroaleman @vincepri hey folks! I updated this PR as you asked - now there is a "featuregate" called `WithExperimentalLazyMapper` that enables the Lazy. PTAL :)

**Review by alvaroaleman — APPROVED** [review]
Thanks you so much!

File: pkg/client/apiutil/lazyrestmapper.go [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+	m.mapper = restmapper.NewDiscoveryRESTMapper(updatedGroupResources)
+
+	return nil
+}
+
+// findAPIGroupByName returns API group by its name.
+func (m *lazyRESTMapper) findAPIGroupByName(groupName string) (metav1.APIGroup, error) {
+	// Ensure that required info about existing API groups is received and stored in the mapper.
+	// It will make 2 API calls to /api and /apis, but only once.
+	if m.apiGroups == nil {
```
Comment (sbueringer) [inline pkg/client/apiutil/lazyrestmapper.go:204]: Sorry very late :)

Q: We only ever seem to write m.apiGroups here. What if the apiGroups change after that? Or do we assume that's not relevant as the CRDs a controller uses must be deployed before the controller?

Comment (sbueringer) [reply pkg/client/apiutil/lazyrestmapper.go:204]: cc @Fedosin

Comment (vincepri) [reply pkg/client/apiutil/lazyrestmapper.go:204]: +1 on the question above, as far as I can see the `apiGroups` is never refreshed which might cause issues for groups that were installed after the mapper gets created AND if someone calls RESTMapping without a version

Comment (fabriziopandini) [reply pkg/client/apiutil/lazyrestmapper.go:204]: Good catch!
Q: does it make sense to track this in an issue and makes sure this has an answer before pushing further the adoption of this mapper?

Comment (vincepri) [reply pkg/client/apiutil/lazyrestmapper.go:204]: Not sure, the code should work for 97% of uses within controller runtime; but the above is definitely an issue we should fix

Comment (sbueringer) [reply pkg/client/apiutil/lazyrestmapper.go:204]: For reference this has been addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/2208 and released in v0.14.5
